heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/HeronMasterDriver.java

```diff
@@ -139,107 +142,129 @@ public HeronMasterDriver(EvaluatorRequestor requestor,
+  boolean scheduleTMasterContainer() {
     LOG.log(Level.INFO, "Scheduling container for TM: {0}", topologyName);
-    try {
-      // TODO: https://github.com/twitter/heron/issues/951: colocate TMaster and Scheduler
-      launchContainerForExecutor(TMASTER_CONTAINER_ID, 1, TM_MEM_SIZE_MB);
-    } catch (InterruptedException e) {
-      // Deployment of topology fails if there is a error starting TMaster
-      throw new RuntimeException("Error while waiting for topology master container allocation", e);
-    }
+    // TODO: https://github.com/twitter/heron/issues/951: colocate TMaster and Scheduler
+    AllocatedEvaluator container =
+        launchContainerForExecutor(new HeronWorker(TMASTER_CONTAINER_ID, 1, TM_MEM_SIZE_MB));
+    return container != null;
   }
 
   /**
    * Container allocation is asynchronous. Request containers serially to ensure allocated resources
    * match the required resources
    */
-  void scheduleHeronWorkers(PackingPlan topologyPacking) {
+  boolean scheduleHeronWorkers(PackingPlan topologyPacking) {
     this.packing = topologyPacking;
     for (Entry<String, ContainerPlan> entry : topologyPacking.containers.entrySet()) {
       Resource reqResource = entry.getValue().resource;
-
       int mem = getMemInMBForExecutor(reqResource);
-      try {
-        launchContainerForExecutor(entry.getKey(), getCpuForExecutor(reqResource), mem);
-      } catch (InterruptedException e) {
-        // Fail deployment of topology if there is a error starting any worker
-        LOG.log(Level.SEVERE, "Container allocation for id:{0} failed. Attempting to close "
-            + "currently allocated resources for {1}", new Object[]{entry.getKey(), topologyName});
-        killTopology();
-        throw new RuntimeException("Failed to allocate container for workers", e);
+      int cores = getCpuForExecutor(reqResource);
+      AllocatedEvaluator container = launchContainerForExecutor(entry.getKey(), cores, mem);
+      if (container == null) {
+        return false;
       }
     }
+    return true;
   }
 
   public void killTopology() {
     LOG.log(Level.INFO, "Kill topology: {0}", topologyName);
-    Map<String, ActiveContext> contexts = clearActiveContexts();
+    isTopologyKilled.set(true);
 
-    for (Entry<String, ActiveContext> entry : contexts.entrySet()) {
-      LOG.log(Level.INFO, "Close context: {0}", entry.getKey());
-      entry.getValue().close();
+    for (HeronWorker worker : multiKeyWorkerMap.getHeronWorkers()) {
+      LOG.log(Level.INFO, "Killing container {0} for executor {1}",
+          new Object[]{worker.evaluator.getId(), worker.workerId});
+      worker.evaluator.close();
     }
   }
 
-  /**
-   * REEF will retain all YARN containers. Currently running heron-executors/REEF-tasks
-   * will be killed and new REEF-tasks representing previously running heron-executor will be
-   * started.
-   */
-  public void restartTopology() {
-    restartContainer(null);
+  public boolean restartTopology() {
```

Comment:
Ugh. Ah yes, the code is riddled with these unfortunate booleans where each callers in the stack checks for false and logs an error and returns false. It's really a poor man's exception throw IMO.

My thought on this is to not propagate this pattern more than we need to. We should implement the boolean return type when the interface demands it, but otherwise we should write new code to expect success and throw an exception on failure. What do you think?